### ISSUE TITLE
In the case of auto cast, turning on activation recomputation, cause OOM

### ISSUE THREAD

**Issue by mars1248** [pytorch/xla#8695]
## To Reproduce

```
import torch
import torch_xla.core.xla_model as xm
import torch_xla.debug.metrics as met
import torch_xla
import torch_xla.utils.checkpoint as checkpoint
from contextlib import nullcontext
from torch_xla.amp import autocast
from torch.nn import Linear
import torch.nn.functional as F
use_autocast = True
#use_autocast = False

class MLPv2(torch.nn.Module):
    def __init__(
        self,
        hidden_size,
        intermediate_size,
        use_bias=False,
    ):

        super().__init__()
        self.hidden_size = hidden_size
        self.intermediate_size = intermediate_size
        self.gate_proj = Linear(self.hidden_size, self.intermediate_size, bias=use_bias)
        self.up_proj = Linear(self.hidden_size, self.intermediate_size, bias=use_bias)
        self.down_proj = Linear(self.intermediate_size, self.hidden_size, bias=use_bias)
        self.act_fn = F.silu

    def forward(self, x):
        ans1 = self.gate_proj(x)
        ans2 = self.up_proj(x)
        ans3 = self.act_fn(ans1)
        ans4 = ans3 * ans2
        down_proj = self.down_proj(ans4)
        return down_proj

def run():
  device = xm.xla_device()
  model = torch.nn.ModuleList([MLPv2(4096, intermediate_size=16384) for _ in range(32)]).to(device)
  if not use_autocast:
    model = model.to(torch.bfloat16)
  optimizer = torch.optim.SGD(model.parameters(), lr=0.0)
  context = nullcontext()
  if use_autocast:
    context = autocast(xm.xla_device(), dtype=torch.bfloat16)
  for step in range(3):
    dummy_data = torch.zeros(1, 4096, 4096, device=device).to(torch.bfloat16)
    optimizer.zero_grad()
    with context:
        x = dummy_data
        for n_l, layer in enumerate(model):
          if n_l > 0:
            x = checkpoint.checkpoint(layer, x)
          else:
            x = layer(x)
        dummy_loss = x.sum()
    dummy_loss.backward()
    optimizer.step()
    # use unlazy to reduce the output size
    xm.unlazy(list(model.parameters())+[dummy_loss])
    #xm.mark_step()


if __name__ == "__main__":
  run()



```

1. run above script
2. use xla autocast need 44GB
3. use bf16 only need 20GB

## Expected behavior
autocast should use similar memory size as bf16

before opt hlo：
```
forward compute:
p92.478 = f32[4096,16384]{1,0} parameter(92), metadata={op_type="xla__device_data" op_name="xla__device_data" source_file="/opt/conda/lib/python3.8/site-packages/torch/nn/modules/module.py" source_line=1326}
transpose.479 = f32[16384,4096]{0,1} transpose(p92.478), dimensions={1,0}, metadata={op_type="aten__permute" op_name="aten__permute" source_file="/opt/conda/lib/python3.8/site-packages/torch/nn/modules/linear.py" source_line=126}
convert.480 = bf16[16384,4096]{0,1} convert(transpose.479), metadata={op_type="xla__cast" op_name="xla__cast" source_file="/opt/conda/lib/python3.8/site-packages/torch/nn/modules/linear.py" source_line=126}
dot.508 = bf16[4096,4096]{1,0} dot(reshape.507, convert.480), lhs_contracting_dims={1}, rhs_contracting_dims={0}, metadata={op_type="aten__mm" op_name="aten__mm" source_file="/opt/conda/lib/python3.8/site-packages/torch/nn/modules/linear.py" source_line=126}
backward compute:
transpose.2899 = f32[16384,4096]{0,1} transpose(p92.478), dimensions={1,0}, metadata={op_type="aten__permute" op_name="aten__permute" source_file="/opt/conda/lib/python3.8/site-packages/torch/nn/modules/linear.py" source_line=126}
convert.2900 = bf16[16384,4096]{0,1} convert(transpose.2899), metadata={op_type="xla__cast" op_name="xla__cast" source_file="/opt/conda/lib/python3.8/site-packages/torch/nn/modules/linear.py" source_line=126}
transpose.2901 = bf16[4096,16384]{1,0} transpose(convert.2900), dimensions={1,0}, metadata={op_type="aten__permute" op_name="aten__permute" source_file="/opt/conda/lib/python3.8/site-packages/torch/autograd/graph.py" source_line=817}
dot.2903 = bf16[4096,16384]{1,0} dot(reshape.2902, transpose.2901), lhs_contracting_dims={1}, rhs_contracting_dims={0}, metadata={op_type="aten__mm" op_name="aten__mm" source_file="/opt/conda/lib/python3.8/site-packages/torch/autograd/graph.py" source_line=817}
```
after opt hlo：
```
forward compute:
wrapped_copy.82 = f32[4096,16384]{1,0} fusion(p38.316.0), kind=kLoop, calls=wrapped_copy_computation.82
wrapped_convert.86 = bf16[4096,16384]{1,0} fusion(wrapped_copy.82), kind=kLoop, calls=wrapped_convert_computation.86, metadata={op_type="xla__cast" op_name="xla__cast" source_file="/opt/conda/lib/python3.8/site-packages/torch/nn/modules/linear.py" source_line=126}
custom-call.428.0 = (bf16[4096,4096]{1,0}, s8[33554432]{0}) custom-call(bitcast.8279.0, wrapped_convert.86), custom_call_target="__cublas$gemm"

backward compute:
get-tuple-element.1194.0 = bf16[1,4096,4096]{2,1,0} get-tuple-element(opt-barrier.1192.0), index=1, metadata={op_type="xla__optimization_barrier" op_name="xla__optimization_barrier" source_file="/opt/conda/lib/python3.8/site-packages/torch_xla/core/xla_model.py" source_line=1495}
bitcast.12492.0 = bf16[4096,4096]{1,0} bitcast(get-tuple-element.1194.0)
custom-call.447.0 = (bf16[4096,16384]{1,0}, s8[33554432]{0}) custom-call(bitcast.12492.0, wrapped_convert.86), custom_call_target="__cublas$gemm"
```
I compared the hlo diagram in the two modes and found that it was due to the mixed precision that the parameter was cast into half precision, and then the parameter of half precision was retained in the backward compute after calculating the forward compute(like above after opt hlo, wrapped_convert.86 was retained). In fact, after calculating the forward direction, it was directly released, and in the reverse direction, it was just cast again(like before opt hlo, p92.478 will cast twice). 
Is there any optimized pass that causes this problem?

## Environment

 - Reproducible on XLA backend [CPU/TPU/CUDA]: CUDA
 - torch_xla version: master

**Comment by ysiraichi** [comment]
Thank you for filing this issue.

I don't think it's necessarily unexpected that AMP would consume more memory than bf16 data-type. The reason for that is that, for example, [there are some operations might cast the inputs to fp32](https://github.com/pytorch/pytorch/blob/6f15a609d395e56ffa433f8c9efa231579bb5fee/aten/src/ATen/autocast_mode.h#L828-L886).

During training, the backend (OpenXLA) is responsible for deciding whether to store the computed values in memory, or to recompute them in the backwards. So, that might be the reason you are seeing that happen in "after opt hlo".

That said, I have 2 questions:
1. Do you observe the same behavior without gradient checkpointing?
2. Do you observe the same behavior using PyTorch?
